### PR TITLE
python 3: Add python 3 support for html tests

### PR DESCRIPTION
**Pull request by ziransun**
[1] Fix some TypeErrors. Calling get API in reuqest.headers
(<class wptserve.request.RequestHeaders>) returns binary string values.
In python 2, bytes is alias to str type. In Python 3, str type is
different from bytes. This change uses utf-8 encoding for the request.headers
values using for Py2 and Py3 compability.
[2] Add python 3 support for accessing headers content in raw_headers
(xhr/resources/inspect-headers.py). Type of raw_headers is class
HTTPMessage. In Python 2, HTTPMessage derives from mimetools.Message while
in Python 3, it derives from email.message.Message. Interfaces from
mimetools.Message and email.message.Message are very different.
In Python 3, HTTPMessage doesn't have 'headers" attribute or equivalent.
[ref: https://bugs.python.org/issue4773]
[3] Fix some imports for python2 and python 3 compatibility

### CODE REVIEW

File: html/semantics/embedded-content/media-elements/track/track-element/cors/support/cors-tester.py [hunk 1]
```diff
@@ -29,11 +28,10 @@ def main(request, response):
         response.headers.set('Access-Control-Allow-Origin', request.GET['origin'])
         response.headers.set('Access-Control-Allow-Credentials', 'true')
 
-    cors = request.headers.get("origin", "no")
-
+    cors = request.headers.get("origin", b"no")
```
Comment (annevk) [inline html/semantics/embedded-content/media-elements/track/track-element/cors/support/cors-tester.py]: I don't understand why this is b"" but true above is not.

File: html/semantics/embedded-content/media-elements/track/track-element/cors/support/cors-tester.py [hunk 2]
```diff
@@ -29,11 +28,10 @@ def main(request, response):
...
     cookie = request.cookies.first(id, None)
     cookie_value = cookie.value if cookie is not None else "no"
-    line = 'cors = ' + cors + ' | cookie = ' + cookie_value;
+    line = 'cors = ' + cors.decode("utf-8") + ' | cookie = ' + cookie_value
```
Comment (annevk) [inline html/semantics/embedded-content/media-elements/track/track-element/cors/support/cors-tester.py]: Is this casting a byte sequence to a string, basically? Are there really UTF-8 bytes in there?

File: html/semantics/embedded-content/the-img-element/404-response-with-actual-image-data.py [hunk 3]
```diff
@@ -1,3 +1,3 @@
 import base64
 def main(req, res):
-    return 404, [('Content-Type', 'image/png')], base64.decodestring("iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsQAAA7EAZUrDhsAAAAhSURBVDhPY3wro/KfgQLABKXJBqMGjBoAAqMGDLwBDAwAEsoCTFWunmQAAAAASUVORK5CYII=")
+    return 404, [('Content-Type', 'image/png')], base64.decodestring(("iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsQAAA7EAZUrDhsAAAAhSURBVDhPY3wro/KfgQLABKXJBqMGjBoAAqMGDLwBDAwAEsoCTFWunmQAAAAASUVORK5CYII=").encode("utf-8"))
```
Comment (annevk) [inline html/semantics/embedded-content/the-img-element/404-response-with-actual-image-data.py]: Why not prefix the string with b""?

Comment (annevk) [reply html/semantics/embedded-content/the-img-element/404-response-with-actual-image-data.py]: Please don't remove the end-of-file newline.

File: xhr/resources/inspect-headers.py [hunk 4]
```diff
@@ -1,9 +1,18 @@
...
-        if line[-2:] != '\r\n':
-            return "Syntax error: missing CRLF: " + line
-        line = line[:-2]
+    if PY3:
+        header_list = (os.linesep.join(
+            [s for s in raw_headers.as_string().splitlines() if s])).split('\n')
+    else:
+        header_list = raw_headers.headers
+    for line in header_list:
+        if not PY3:
```
Comment (annevk) [inline xhr/resources/inspect-headers.py]: This seems wrong.

**Review by annevk — COMMENTED** [review]
I'm not a big fan of all the encode/decode operations here. There might be a better way to accomplish casting from one type to another.

File: xhr/resources/inspect-headers.py (hunk 4 above)
Comment (ziransun) [reply xhr/resources/inspect-headers.py]: Is it okay to be more specific? raw_headers derives from different classes in Python 2 and Python 3. Here is to handle them differently.

**Comment by ziransun** [comment]
Code has been updated as per review comments. Please review. Thanks!

File: html/semantics/forms/form-submission-0/resources/form-submission.py [hunk 5]
```diff
@@ -1,8 +1,9 @@
+from six import ensure_str
 def main(request, response):
-    if request.headers.get('Content-Type') == 'application/x-www-form-urlencoded':
-        result = request.body == 'foo=bara'
-    elif request.headers.get('Content-Type') == 'text/plain':
-        result = request.body == 'qux=baz\r\n'
+    if ensure_str(request.headers.get('Content-Type')) == 'application/x-www-form-urlencoded':
```
Comment (annevk) [inline html/semantics/forms/form-submission-0/resources/form-submission.py]: Wouldn't it be easier to use b"" here (and below) on the other side?

File: tools/wptserve/wptserve/request.py [hunk 6]
```diff
@@ -1,9 +1,9 @@
 import base64
 import cgi
-from six.moves.http_cookies import BaseCookie
-from six import BytesIO, binary_type, text_type, iteritems, PY3
 import tempfile
 
+from six import BytesIO, binary_type, ensure_str, text_type, iteritems, PY3
+from six.moves.http_cookies import BaseCookie
```
Comment (annevk) [inline tools/wptserve/wptserve/request.py]: Why was this moved? @jgraham should probably look at this change.

File: xhr/resources/inspect-headers.py (hunk 4 above)
Comment (annevk) [reply xhr/resources/inspect-headers.py]: In particular, it seems like the code isn't executed at all in Python 3.

Comment (ziransun) [reply xhr/resources/inspect-headers.py]: raw_headers derives from different classes in python 2 and python 3. In Python 3, HTTPMessage doesn't have 'headers" attribute or equivalent as we described in the commit message (see the reference as well). That's why this part code is not executed in Python 3.

File: tools/wptserve/wptserve/request.py (hunk 6 above)
Comment (ziransun) [reply tools/wptserve/wptserve/request.py]: Reordered these lines due to pylint complains.

File: html/semantics/forms/form-submission-0/resources/form-submission.py (hunk 5 above)
Comment (ziransun) [reply html/semantics/forms/form-submission-0/resources/form-submission.py]: Shouldn't this kind information be string?

Comment (annevk) [reply html/semantics/forms/form-submission-0/resources/form-submission.py]: Why?

File: xhr/resources/inspect-headers.py (hunk 4 above)
Comment (annevk) [reply xhr/resources/inspect-headers.py]: We can't just drop testing coverage. We probably need a better solution to how we deal with header logic migration in that case.

Comment (ziransun) [reply xhr/resources/inspect-headers.py]: Here a few new lines added for python 3 to perform similar functionality. It basically hasn't touched/removed any original python 2 code.

Comment (annevk) [reply xhr/resources/inspect-headers.py]: Where does the Python 3 code test for a missing CRLF?

**Comment by jgraham** [comment]
I think per the decision not to work on converting handlers until we have a better testing story, we should close this one for now. Feel free to reopen if I got that wrong.
